internal/llmcostcel/cel.go

```diff
@@ -0,0 +1,79 @@
+// Package llmcostcel provides functions to create and evaluate CEL programs to calculate costs.
+//
+// This exists as a separate package to be used both in the controller to validate the expression
+// and in the external processor to evaluate the expression.
+package llmcostcel
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+)
+
+const (
+	celModelNameKey    = "model"
+	celBackendKey      = "backend"
+	celInputTokensKey  = "input_tokens"
+	celOutputTokensKey = "output_tokens"
+	celTotalTokensKey  = "total_tokens"
+)
+
+// NewProgram creates a new CEL program from the given expression.
+func NewProgram(expr string) (prog cel.Program, err error) {
+	env, err := cel.NewEnv(
+		cel.Variable(celModelNameKey, cel.StringType),
+		cel.Variable(celBackendKey, cel.StringType),
+		cel.Variable(celInputTokensKey, cel.UintType),
+		cel.Variable(celOutputTokensKey, cel.UintType),
+		cel.Variable(celTotalTokensKey, cel.UintType),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create CEL environment: %w", err)
+	}
+	ast, issues := env.Compile(expr)
+	if issues != nil && issues.Err() != nil {
+		err := issues.Err()
+		return nil, fmt.Errorf("cannot compile CEL expression: %w", err)
+	}
+	prog, err = env.Program(ast)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create CEL program: %w", err)
+	}
+
+	// Sanity check by evaluating the expression with some dummy values.
+	_, err = EvaluateProgram(prog, "dummy", "dummy", 0, 0, 0)
```

Comment:
This seems like getting called on the data path as well for each request, can we only initialize once?

Reply:
`prog` is initialized on the config load time which is the compiled CEL program. so i don't think we could do beyond that (the other args are depending on the request context)

Reply:
btw to be clear, this function itself `NewProgram` is only called on config path and `prog` is initialized once here. And then on the request path only `EvaluateProgram` is called using the precompiled `prog`

Reply:
You are right I read the code wrong.